node-it/src/test/scala/com/wavesplatform/it/sync/AddressApiSuite.scala

```diff
@@ -33,10 +46,89 @@ class AddressApiSuite extends BaseTransactionSuite {
         sender.getData(firstAddress, invalidRegexp)
         fail("RegexCompiler didn't throw expected error")
       } catch {
-        case err: Throwable => if (!err.getMessage.contains("Cannot compile regex")) {
-          throw err
-        }
+        case err: Throwable =>
+          if (!err.getMessage.contains("Cannot compile regex")) {
+            throw err
+          }
       }
     }
   }
+
+  test("balances for waves should be correct") {
+    assertBalances(None)
+  }
+
+  test("balances for issued asset should be correct") {
+    val asset = miner.issue(miner.address, "Test", "Test", 10000, 0, waitForTx = true).id
+    assertBalances(Some(asset))
+  }
+
+  test("limit violation requests should be handled") {
+    val limit     = miner.config.getInt("waves.rest-api.transactions-by-address-limit")
+    val address   = miner.createAddress()
+    val addresses = List.fill(limit + 1)(address)
+    assertApiError(
+      miner.get(s"/addresses/balance?${addresses.map(a => s"address=$a").mkString("&")}"),
+      TooBigArrayAllocation
+    )
+    assertApiError(
+      miner.postJson(
+        "/addresses/balance",
+        Json.obj("ids" -> addresses)
+      ),
+      TooBigArrayAllocation
+    )
+  }
+
+  private def assertBalances(asset: Option[String]): Unit = {
+    val addresses = (1 to 5).map(i => (miner.createAddress(), (i * 100).toLong)).toList
+
+    val firstAddresses  = addresses.slice(0, 2)
+    val secondAddresses = addresses.slice(2, 5)
+
+    val heightBefore  = transferAndReturnHeights(firstAddresses, asset).min - 1
+    val heightBetween = nodes.waitForHeightArise()
+    nodes.waitForHeightArise() // prevents next transfers from accepting on the heightBetween
+    val heightAfter = transferAndReturnHeights(secondAddresses, asset).max
+
+    nodes.waitForHeightArise()
+
+    checkBalances(List(), addresses.map(_._1), Some(heightBefore), asset)          // balances at the height before all transfers
+    checkBalances(firstAddresses, addresses.map(_._1), Some(heightBetween), asset) // balances at the height after the 2nd transfer
+    checkBalances(addresses, addresses.map(_._1), Some(heightAfter), asset)        // balances at the height after all transfers
+    checkBalances(addresses, addresses.map(_._1), None, asset)                     // balances at the current height
+  }
+
+  private def transferAndReturnHeights(addresses: List[(String, Long)], asset: Option[String]): List[Int] = {
+    val ids = addresses.map { case (address, a) => miner.transfer(miner.address, address, a, minFee, asset).id }
+    ids.map(id => miner.waitForTransaction(id).height)
+  }
+
+  private def checkBalances(expected: List[(String, Long)], addresses: List[String], height: Option[Int], assetId: Option[String]): Unit = {
+    val getResult = miner.get(
+      s"/addresses/balance?${height.fold("")(h => s"height=$h&")}${assetId.fold("")(a => s"asset=$a&")}${addresses.map(a => s"address=$a").mkString("&")}"
```

Comment:
move endpoint call to SyncHttpApi and AsyncHttpApi to use these methods in other tests

Reply:
`POST` version added to `HttpApi`